daml-lf/interface/src/main/scala/com/digitalasset/daml/lf/iface/DefDataType.scala

```diff
@@ -173,11 +173,30 @@ final case class Enum(constructors: ImmArraySeq[Ref.Name]) extends DataType[Noth
   def asDataType[RT, PVT]: DataType[RT, PVT] = this
 }
 
-final case class DefTemplate[+Ty](choices: Map[Ref.Name, TemplateChoice[Ty]], key: Option[Ty]) {
+final case class DefTemplate[+Ty](
+    choices: Map[Ref.ChoiceName, TemplateChoice[Ty]],
+    inheritedChoices: Map[Ref.ChoiceName, Ref.TypeConName],
+    key: Option[Ty],
+) extends DefTemplate.GetChoices[Ty] {
   def map[B](f: Ty => B): DefTemplate[B] = Functor[DefTemplate].map(this)(f)
 
-  def getChoices: j.Map[Ref.ChoiceName, _ <: TemplateChoice[Ty]] =
-    choices.asJava
+  /** Remove choices from `inheritedChoices` and add to `choices` given the
+    * `astInterfaces` from an [[EnvironmentInterface]].  If the result has any
+    * `inheritedChoices` left, these choices were not found.
+    */
+  def resolveChoices[O >: Ty](
```

Comment:
Possibly should be private as a utility for the function of same name in `EnvironmentInterface`.

Reply:
I agree, however I think this could solve itself with resolving the issue that `inheritedChoices` should actually be empty in the end for the user.

Reply:
It became apparent looking at #12689 that having `EnvironmentInterface` only as an entry point would be really inconvenient, hence the profusion of additional accessible utilities in this PR.